File: lab2/gr8sub1/ГУЗОВА_АННА/src/ptree.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

static void print_chain(pid_t pid) {
    if (pid == 0) return;

    char path[64];
    snprintf(path, sizeof(path), "/proc/%d/status", pid);
    FILE *f = fopen(path, "r");
    if (!f) return;

    char *line = NULL;
    size_t len = 0;
    pid_t ppid = 0;
    char comm[256] = "unknown";

    while (getline(&line, &len, f) != -1) {
        if (strncmp(line, "Name:", 5) == 0) {
            sscanf(line + 5, "%255s", comm);
        }
        if (strncmp(line, "PPid:", 5) == 0) {
            sscanf(line + 5, "%d", &ppid);
        }
    }
    free(line);
    fclose(f);

    if (ppid != 0) {
        print_chain(ppid);
        printf(" ← ");
    }
    printf("%s(%d)", comm, pid);
}

int main(void) {
    print_chain(getpid());
    putchar('\n');
    return 0;
}

File: lab2/gr9sub2/Butrameev_Ivan/src/ptree.py
import os

pid = os.getpid()
chain = []

while pid > 0:
    name, ppid = None, 0
    with open(f"/proc/{pid}/status") as f:
        for line in f:
            if line.startswith("Name:"):
                name = line.split()[1]
            elif line.startswith("PPid:"):
                ppid = int(line.split()[1])
    chain.append(f"{name}({pid})")
    pid = ppid

print(" ← ".join(chain))
